src/Build/Construction/ProjectRootElement.cs

```diff
@@ -1584,6 +1586,7 @@ public void Save(TextWriter writer)
                 XmlDocument.Save(projectWriter);
             }
 
+            StreamTime = DateTime.Now;
```

Comment:
Is this guaranteed to write to disk and to the same path that represents the file?  Why not just update the last write time?  If its a memory stream, or anything not on disk, then it wouldn't trigger a rebuild anyway since the datetime stamp would be calculated off of the file during build.  So it seems like it would be better to just assume the file was updated on disk if the stream was updated so everything works the same.  Unless of course I'm missing something.

Reply:
The "last modified project" [is calculated](https://github.com/microsoft/msbuild/blob/master/src/Build/Evaluation/Evaluator.cs#L1986-L1989) purely based on `_lastWriteTimeWhenRead`. Since we only keep the last modified project to add to MSBuildAllProjects, it would break that should the last thing we wrote to not be a file on disk.

Reply:
> Since we only keep the last modified project to add to MSBuildAllProjects, it would break that should the last thing we wrote to not be a file on disk.

I'm still confused.  If we didn't write a file to disk, the up-to-date check wouldn't pick it up either right?  Just adding the path of the import/project to MSBuildAllProjects isn't enough, the timestamp on disk has to change as well.

Maybe we just need a new LastWriteTime property which reflects when the file is modified but doesn't cause the file to be re-read from disk?  And then this calculation would come from that.

Reply:
This is the scenario I mean:
```
Write to file a
Build
Write to file b
Write to stream c that is not represented by a file on disk
Build
```
In this case, if we update the last write time of the projectrootelement corresponding to c, then in the part of Evaluator I linked above, it will think the last modified project was c, not b, so it will prepend that to MSBuildAllProjects.
When we check to see if we need to rebuild, there's no corresponding file on disk for the last project "written to," so since we only kept the last project (for performance) instead of keeping all of them, our up-to-date check fails—we miss that file b was written to.
In other words, the problem is that when we write to a stream, it could be writing to a file (in which case, we should update _lastWriteTimeWhenRead) or not, in which case doing so would prevent us from noticing the real most recently written ProjectRootElement.

Could you explain your LastWriteTime idea a little more and how we can detect that a file has been modified without reading it? I don't understand how it differs from _lastWriteTimeWhenRead.